js/components/authoring.js

```diff
@@ -17,104 +17,129 @@ const AVAILABLE_PRESETS = [
     const { step } = this.state
-    return step === 4 ? 'finish' : 'next'
+    return config.authoringSteps[step]
   }
 
-  get buttonDisabled () {
+  get nextButtonLabel () {
     const { step } = this.state
-    return step === 1
+    return step === config.authoringSteps.length - 1 ? 'finish' : 'next'
   }
 
-  handleNextButtonClick () {
+  get navigationDisabled () {
+    return this.currentStep === 'presets'
+  }
+
+  componentDidUpdate (prevProps, prevState) {
     const { step } = this.state
-    const { setOption } = this.props
-    if (step === 2) {
-      this.snapshotStep()
-      this.setStep3()
-    } else if (step === 3) {
-      if (config.densityStepEnabled) {
-        this.snapshotStep()
-        setOption('interaction', 'none')
-        setOption('selectableInteractions', [])
-        setOption('colormap', 'plate')
-        this.setState({step: 4})
-      } else {
+    if (step !== prevState.step) {
+      const stepName = this.currentStep
+      if (stepName === 'presets') {
+        this.unloadModel()
+      } else if (stepName === 'continents') {
+        this.setContinentsStep()
+      } else if (stepName === 'forces') {
+        this.setForcesStep()
+      } else if (stepName === 'densities') {
+        this.setDensitiesStep()
+      } else if (stepName === undefined) {
         this.endAuthoring()
       }
-    } else if (step === 4) {
-      this.endAuthoring()
+
+      if (step > prevState.step) {
+        this.saveModel()
+      } else if (stepName !== 'presets') {
+        this.restoreModel()
+      }
     }
   }
 
+  handleNextButtonClick () {
+    const { step } = this.state
+    this.setState({ step: step + 1 })
+  }
+
   handleBackButtonClick () {
     const { step } = this.state
-    const { restoreLabeledSnapshot } = this.props
-    if (step === 2) {
-      this.unloadModel()
-    } else if (step === 3) {
-      restoreLabeledSnapshot(STEP_NAMES[2])
-      this.setStep2()
-    } else if (step === 4) {
-      restoreLabeledSnapshot(STEP_NAMES[3])
-      this.setStep3()
+    if (step > 0) {
+      this.setState({ step: step - 1 })
     }
   }
 
-  handleInteractionChange (interactionName) {
-    const { setOption } = this.props
-    setOption('interaction', interactionName)
+  saveModel () {
+    const { takeLabeledSnapshot } = this.props
+    takeLabeledSnapshot(this.currentStep)
   }
 
-  snapshotStep () {
-    const { step } = this.state
-    const { takeLabeledSnapshot } = this.props
-    takeLabeledSnapshot(STEP_NAMES[step])
+  restoreModel () {
+    const { restoreLabeledSnapshot } = this.props
+    restoreLabeledSnapshot(this.currentStep)
   }
 
   loadModel (presetInfo) {
     const { loadModel } = this.props
-    loadModel(presetInfo.name)
-    this.setStep2()
+    this.setState({ modelLoading: true })
+    loadModel(presetInfo.name, () => {
+      // Important! Go to the next step after model load is complete.
+      // Otherwise, we might try to take model snapshot too early (when it's not loaded yet).
```

Comment:
Nice fix, this definitely wasn't addressed before

Reply:
Implementation has changed in the following commits and now it's addressed using a message queue. Risk of reviewing commit by commit. 😉